test/Concurrency/Runtime/async_stream.swift

```diff
@@ -415,6 +415,58 @@ class NotSendable {}
         expectTrue(expectation.fulfilled)
       }
 
+      tests.test("finish() behavior with multiple values being awaited") {
+        let expectationCount = 3
+        let e1 = Expectation()
+        let e2 = Expectation()
+        let e3 = Expectation()
+
+        func getExpectation(for index: Int) -> Expectation {
+          switch index {
+            case 1: return e1
+            case 2: return e2
+            case 3: return e3
+            default: preconditionFailure("invalid index")
+          }
+        }
+
+        let (stream, continuation) = AsyncStream<Int>.makeStream()
+        let (controlStream, controlContinuation) = AsyncStream<Int>.makeStream()
+
+        await withTaskGroup(of: Void.self) { group in
+          for i in 1...expectationCount {
+            group.addTask {
+              for await _ in stream {
+                controlContinuation.yield(i)
+              }
+              getExpectation(for: i).fulfilled = true
+            }
+          }
+
+          // Allow consumers to progress.
```

Comment:
i think there's still some non-determinism here... the setup we want to exercise is when there are `n` pending continuations awaiting values from the underlying stream. i don't think this setup guarantees that actually occurs. but anecdotally, this type of test would never complete under the prior stream implementation.

Reply:
pretty sure the non-determinism is now resolved, and we should be guaranteed the test is in a state with 2 pending continuations when the stream terminates.